pkg/controller/keyvault/nginx_secret_provider_class.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/controllername"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/metrics"
+	"github.com/Azure/aks-app-routing-operator/pkg/manifests"
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	kvcsi "github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/provider/types"
+)
+
+var (
+	nginxSecretProviderControllerName = controllername.New("keyvault", "nginx", "secret", "provider")
+	NginxNamePrefix                   = "keyvault-nginx-"
+)
+
+// NginxSecretProviderClassReconciler manages a SecretProviderClass for each nginx ingress controller that
+// has a Keyvault URI in its DefaultSSLCertificate field. The SPC is used to mirror the Keyvault values into
+// a k8s secret so that it can be used by the CRD controller.
+type NginxSecretProviderClassReconciler struct {
+	client client.Client
+	events record.EventRecorder
+	config *config.Config
+}
+
+func NewNginxSecretProviderClassReconciler(manager ctrl.Manager, conf *config.Config) error {
+	metrics.InitControllerMetrics(nginxSecretProviderControllerName)
+	if conf.DisableKeyvault {
+		return nil
+	}
+	return nginxSecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&approutingv1alpha1.NginxIngressController{}), manager.GetLogger(),
+	).Complete(&NginxSecretProviderClassReconciler{
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	})
+}
+
+func (i *NginxSecretProviderClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var err error
+	result := ctrl.Result{}
+
+	// do metrics
+	defer func() {
+		//placing this call inside a closure allows for result and err to be bound after Reconcile executes
+		//this makes sure they have the proper value
+		//just calling defer metrics.HandleControllerReconcileMetrics(controllerName, result, err) would bind
+		//the values of result and err to their zero values, since they were just instantiated
+		metrics.HandleControllerReconcileMetrics(nginxSecretProviderControllerName, result, err)
+	}()
+
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return result, err
+	}
+	logger = nginxSecretProviderControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	logger.Info("getting Nginx Ingress")
+	nic := &approutingv1alpha1.NginxIngressController{}
+	err = i.client.Get(ctx, req.NamespacedName, nic)
+	if err != nil {
+		return result, client.IgnoreNotFound(err)
+	}
+	logger = logger.WithValues("name", nic.Name, "namespace", "app-routing-system", "generation", nic.Generation)
+
+	spc := &secv1.SecretProviderClass{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "secrets-store.csi.x-k8s.io/v1",
+			Kind:       "SecretProviderClass",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      DefaultNginxCertName(nic),
+			Namespace: "app-routing-system",
```

Comment:
we should be taking the namespace through the config. https://github.com/Azure/aks-app-routing-operator/blob/main/pkg/config/config.go

example of another controller doing this https://github.com/Azure/aks-app-routing-operator/blob/aa4f99d481269d43a42df4f9e8ee785b60ab394c/pkg/controller/nginxingress/nginx_ingress_controller.go#L51